kubernetes/controller/internal/resolution/cached_repository.go

```diff
@@ -0,0 +1,127 @@
+		repo:       repo,
+	}
+}
+
+// AddComponentVersion adds a component version to the underlying repository.
+func (c *CacheBackedRepository) AddComponentVersion(ctx context.Context, descriptor *descriptor.Descriptor) error {
+	return c.repo.AddComponentVersion(ctx, descriptor)
+}
+
+// GetComponentVersion retrieves a component version, using the cache when possible.
+func (c *CacheBackedRepository) GetComponentVersion(ctx context.Context, component, version string) (*descriptor.Descriptor, error) {
+	var configHash []byte
+	if c.cfg != nil {
+		configHash = c.cfg.Hash
+	}
+
+	key, err := buildCacheKey(configHash, c.spec, component, version)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build cache key: %w", err)
+	}
+
+	wpOpts := workerpool.ResolveOptions{
+		RepositorySpec: c.spec,
+		Component:      component,
+		Version:        version,
+		Repository:     c.repo,
+		Key:            key,
+	}
+
+	result, err := c.workerPool.GetComponentVersion(ctx, wpOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// ListComponentVersions lists all versions of a component, using the cache when possible.
+func (c *CacheBackedRepository) ListComponentVersions(ctx context.Context, component string) ([]string, error) {
+	return c.repo.ListComponentVersions(ctx, component)
+}
+
+// AddLocalResource adds a local resource to the underlying repository.
+// TODO: Eventually cache these as well.
+func (c *CacheBackedRepository) AddLocalResource(ctx context.Context, component, version string, res *descriptor.Resource, content blob.ReadOnlyBlob) (*descriptor.Resource, error) {
+	return c.repo.AddLocalResource(ctx, component, version, res, content)
+}
+
+// GetLocalResource retrieves a local resource from the underlying repository.
+func (c *CacheBackedRepository) GetLocalResource(ctx context.Context, component, version string, identity runtime.Identity) (blob.ReadOnlyBlob, *descriptor.Resource, error) {
+	return c.repo.GetLocalResource(ctx, component, version, identity)
+}
+
+// AddLocalSource adds a local source to the underlying repository.
+func (c *CacheBackedRepository) AddLocalSource(ctx context.Context, component, version string, src *descriptor.Source, content blob.ReadOnlyBlob) (*descriptor.Source, error) {
+	return c.repo.AddLocalSource(ctx, component, version, src, content)
+}
+
+// GetLocalSource retrieves a local source from the underlying repository.
+func (c *CacheBackedRepository) GetLocalSource(ctx context.Context, component, version string, identity runtime.Identity) (blob.ReadOnlyBlob, *descriptor.Source, error) {
+	return c.repo.GetLocalSource(ctx, component, version, identity)
+}
+
+// CheckHealth calls health check on the underlying repository. Returns an error if the repository does not support
+// health checking.
+func (c *CacheBackedRepository) CheckHealth(ctx context.Context) error {
+	checkable, ok := c.repo.(repository.HealthCheckable)
+	if !ok {
+		return fmt.Errorf("repository is not healthy")
```

Comment:
I dont think this is the correct behavior. Instead there should probably be a log and a noop. Just because its not checkable doesnt mean its not healthy. I.e. CTFs